### DISCUSSION TITLE
Streamlining DLL/so library inclusion for CQ-Editor

### DISCUSSION THREAD

**Discussion by jdegenstein** [pyinstaller/pyinstaller#7199] [Help]
I have been working on pyinstaller static builds for my fork of an open source CodeCAD editor called CQ-Editor. I recently discovered that one of the features was broken because it was missing a DLL/so (turned out it was missing several). The errors took a while to work through, but I was able to figure out what libraries I was missing by using procmon. I now have a working pyinstaller build but I have to specify the missing DLLs one-by-one, which feels fragile in case any of the dependencies change in the future. My fork of CQ-editor is located [here](https://github.com/jdegenstein/jmwright-CQ-Editor) along with the pyinstaller.spec and github actions.

Here are the DLLs I had to copy (using binaries in my pyinstaller.spec):
```
casadi_nlpsol_ipopt.dll
ipopt-3.dll
libblas.dll
liblapack.dll
dmumps.dll
flang.dll
flangrti.dll
libomp.dll
mkl_core.2.dll
mkl_def.2.dll
mkl_intel_thread.2.dll
mkl_msg.dll
#currently just including all mkl*.dll
```
What is the preferred way of avoiding this situation? Do I need to write a pyinstaller hook, or is there something I am missing?

**Comment by bwoodsend** [comment]
[Sigh] And this is why we keep saying that PyInstaller should never be used with Anaconda...

**Reply by jdegenstein** [reply]
Hmm, ok so it is always preferred to use pip? Sorry I am very new to all of this. I have also been working on getting pip running for CQ-editor, so maybe this will be the motivation I need to finish it.

**Reply by bwoodsend** [reply]
Yes, stick to pip. Anaconda rearranges the contents of packages into something resembling a UNIX style environment which, as I understand it, is only possible on Windows by _weak linking_ Python extension modules to the DLLs that they depend on. PyInstaller cannot reverse engineer that.

**Reply by jdegenstein** [reply]
Hello, I was able to get pip working but I am still facing the same issue with missing libraries. I start by creating a new python 3.9 environment in micromamba and then all packages are installed using pip. See here:

https://github.com/jdegenstein/jmwright-CQ-Editor/blob/main/.github/workflows/pyinstaller-builds-actions-PIP-TAR.yml#L102

It looks like I am missing the same libraries as before. Any advice?

**Reply by rokm** [reply]
> Hello, I was able to get pip working but I am still facing the same issue with missing libraries. I start by creating a new python 3.9 environment in micromamba and then all packages are installed using pip. See here:
> 
> https://github.com/jdegenstein/jmwright-CQ-Editor/blob/main/.github/workflows/pyinstaller-builds-actions-PIP-TAR.yml#L102
> 
> It looks like I am missing the same libraries as before. Any advice?

You're explicitly installing non-standard `numpy` (via `pipwin install numpy`), of which PyInstaller's numpy hook has no awareness and hence no support for. So if you really want to use MKL-enabled `numpy` (with weakly-linked dependencies), you'll need to collect those DLLs yourself in the .spec file (in this case, there's likely no benefit of doing it in a hook vs. spec file, unless perhaps if you wanted to reuse the hook in multiple projects).

**Reply by rokm** [reply]
And make sure you really collect all of them, because if I recall correctly, the whole point of weak-linking/dynamic DLL loading in numpy+MKL is that DLLs are loaded based on CPU arch and capabilities...

**Reply by jdegenstein** [reply]
Thank you very much for the reply. I have run a new build after removing the pip install of pipwin and then "pipwin install numpy". This greatly reduced the number of missing libraries, but it was still missing a few. I tracked them down using procmon as before and manually copied into the build and got CQ-editor working. Here are the libraries I had to manually copy:
```
libcasadi_nlpsol_ipopt.dll
libgfortran-3.dll
libquadmath-0.dll
```
Is there any way to automatically include these?

EDIT: I copied these DLLs from the PyPi version of casadi==3.5.5

**Reply by rokm** [reply]
Those are probably also dynamically loaded, so they cannot be automatically picked up during binary dependency analysis.

```python
from PyInstaller.utils.hooks import collect_dynamic_libs
binaries = collect_dynamic_libs('casadi')
```

You can add that to your spec file (and pass contents of `binaries`) to `Analysis`.

Or you could store that snippet to `hook-casadi.py` and pass the path to the hook's parent directory via `hookspath` argument to `Analysis` in the spec.

EDIT: that will collect *all* of them - if you want collection to be limited to just the ones you listed, you'll need to filter the returned `binaries` list.

**Reply by rokm** [reply]
That said, `casadi` might be doing some `sys.path` manipulation that causes `casadi._casadi` extension to be collected as a top-level `_casadi` module; and since it ends up in top-level application directory and not next to the DLLs in the `casadi` sub-directory, it might fail to find them...

**Reply by rokm** [reply]
> That said, `casadi` might be doing some `sys.path` manipulation that causes `casadi._casadi` extension to be collected as a top-level `_casadi` module; and since it ends up in top-level application directory and not next to the DLLs in the `casadi` sub-directory, it might fail to find them...

Or rather, something about the legacy import codepaths in `casadi.casadi` causes our analysis to pick `casadi._casadi` as top-level `_casadi`. The following hook makes a simple `import casadi` program run for me in frozen form (latest PyInstaller version):

```python
# hook-casadi.py
from PyInstaller.utils.hooks import collect_dynamic_libs

binaries = collect_dynamic_libs('casadi')

# Something about legacy import codepaths in casadi.casadi causes PyInstaller's analysis to pick up
# casadi._casadi as a top-level _casadi module, which is wrong.
hiddenimports = ['casadi._casadi']
excludedimports = ['_casadi']
```

You can probably also pass equivalent `hiddenimports` and `excludes` to Analysis in the spec file.

**Reply by jdegenstein** [reply]
Thanks for the recommendations, I am trying your `hook-casadi.py` file now in a new build. I am thinking the way that `cadquery/occ_impl/solver.py` imports casadi may also have something to do with it, the necessary DLLs aren't "decided" until a later call to `casadi.Opti().solver()`.

**Reply by jdegenstein** [reply]
So that did not seem to work, no change from the prior run as far as any libraries are concerned. I included this line in `pyinstaller_pip.spec`:
```py
hookspath=['pyinstaller/hook-casadi.py'],
```

**Reply by rokm** [reply]
You need to pass path to extra hooks directory via `hookspath` (as opposed to the path to the hooks file).

**Reply by rokm** [reply]
And you'll know that hook is actually ran when you see a line about the `hook-casadi.py` being ran in your build log, similar to these:
https://github.com/jdegenstein/jmwright-CQ-Editor/actions/runs/3346572413/jobs/5543540667#step:5:105

**Reply by jdegenstein** [reply]
That worked! Thank you very much for all your help. It included all the casadi libraries, but that is OK for now.
